apache2/waf_logging/waf_log_util.cc

```diff
@@ -54,12 +55,63 @@ static std::string to_json_string(const waf_logging::waf_log& log) {
     return json_output;
 };
 
-static std::vector<int> default_disableable_rule_ids;
+enum class rule_set_types {
+    TYPE_UNKNOWN,
+    OWASP_CRS,
+};
+
+enum class rule_set_versions {
+    VERSION_UNKNOWN,
+    VERSION_2_2_9,
+    VERSION_3_0,
+    VERSION_3_1,
+};
+
+using rule_set_info = std::pair<rule_set_types, rule_set_versions>;
+static const rule_set_info OWASP_2_2_9 {std::make_pair(rule_set_types::OWASP_CRS, rule_set_versions::VERSION_2_2_9)};
+static const rule_set_info OWASP_3_0 {std::make_pair(rule_set_types::OWASP_CRS, rule_set_versions::VERSION_3_0)};
+static const rule_set_info OWASP_3_1 {std::make_pair(rule_set_types::OWASP_CRS, rule_set_versions::VERSION_3_1)};
+
+static rule_set_types parse_rule_set_type(const char* type) {
+    if (!type) {
+        return rule_set_types::TYPE_UNKNOWN;
+    }
+
+    return (std::strcmp(type, "OWASP_CRS") == 0) ? rule_set_types::OWASP_CRS : rule_set_types::TYPE_UNKNOWN;
+}
+
+static rule_set_versions parse_rule_set_version(const char* version) {
+    if (!version) {
+        return rule_set_versions::VERSION_UNKNOWN;
+    }
+
+    return (std::strcmp(version, "2.2.9") == 0) ? rule_set_versions::VERSION_2_2_9 :
+        (std::strcmp(version, "3.0.0") == 0) ? rule_set_versions::VERSION_3_0 :
+        (std::strcmp(version, "3.1.0") == 0) ? rule_set_versions::VERSION_3_1 :
+        rule_set_versions::VERSION_UNKNOWN;
+}
+
+using rule_ids_vector = std::vector<int>;
```

Comment:
Use a set instead?

Reply:
For the given use case, a sorted vector is better than a set for multiple reasons:
1) It is more cache-friendly for a vector inherently has the best cache locality among common data structures but a set is effectively a Red-Black tree which relies on pointers that may stand far from each other.
2) Vector is more memory-friendly for the same reason. A set involves extra pointers for each node in addition to the value it stores.

A set can give an edge for the case when inserts and lookups are interleaved because in this scenario vector has poor performance on inserting elements in its middle while keeping it sorted. In our case we only read (e.g., insert everything) once in the beginning and only perform lookups. A sorted vector is a perfect fit for this.

You can find more on the subject in this paper: http://lafstern.org/matt/col1.pdf

Reply:
Agreed that writes would make a better case for a set.